### ISSUE TITLE
Kubernetes agents stop polling after a while

### ISSUE THREAD

**Issue by aufarg** [spinnaker/spinnaker#6288]
### Issue Summary:

Clouddriver Kubernetes agents doesn't poll for new updates after running for some time.

### Cloud Provider(s):

Kubernetes

### Environment:

Spinnaker 1.22.4, deployed on AWS ECS

### Feature Area:

Kubernetes, Clouddriver

### Description:

We have several cases where some of the agents polling from Kubernetes stops polling after running perfectly fine for several days. This makes updates on Kubernetes resources to not be reflected in the UI.

Restarting Clouddriver seems to make the agents poll again.

### Steps to Reproduce:

Not quite sure how this happens, but we have several clusters registered on our Clouddriver with Dual SQL + Redis backend configured as follow:

```
sql:
  enabled: true
  read-only: false
  taskRepository:
    enabled: true
  cache:
    enabled: true
    readBatchSize: 500
    writeBatchSize: 300
  scheduler:
    enabled: false
  unknown-agent-cleanup-agent:
    enabled: false

redis:
  scheduler: default
  parallelism: -1
  configuration:
    secure: true
  cache:
    enabled: false
  taskRepository:
    enabled: true
```

### Additional Details:

After investigating the log, we noticed that when the agent is running correctly, it is supposed to omit logs with patterns like:
```
KubernetesCoreCachingAgent[1/1]: agent is starting
KubernetesCoreCachingAgent[1/1]: grouping <resource> has 4078 entries and 4075 relationships
... <several lines of log like above with different resources> ...
```
The last log seen from the agent tho, only logs the first line:
```
KubernetesCoreCachingAgent[1/1]: agent is starting
```
After that, there's no more logs from the agent.

To me, this looks like the agent does start (so not the problem on scheduling), but stuck while trying to do its task.

From the code, the grouping logs comes from [here](https://github.com/spinnaker/clouddriver/blob/2a00ac45e4f732917d50b444b7b37844391aa352/clouddriver-kubernetes/src/main/java/com/netflix/spinnaker/clouddriver/kubernetes/caching/agent/KubernetesV2CachingAgent.java#L205) (specifically `logStratifiedCacheData`). That means somewhere along these [lines](https://github.com/spinnaker/clouddriver/blob/2a00ac45e4f732917d50b444b7b37844391aa352/clouddriver-kubernetes/src/main/java/com/netflix/spinnaker/clouddriver/kubernetes/caching/agent/KubernetesV2OnDemandCachingAgent.java#L78-L114), the agent gets blocked.

---

**Comment by clanesf** [comment]
We've had problems similar to this when running the AgentScheduler on SQL and not on Redis, though from your configuration it seems like you are running the AgentScheduler on Redis, Can you confirm this?

You may also try turning debugging logging on for the CachingAgents to see if there some kind of error being swallowed. I've seen similar issues when Spinnaker tries to read a kind that doesn't exist on the target cluster (specifically the podPreset) kind. In that case I had to add `podPreset` to omitKinds to get it to cache the cluster correctly.

**Comment by aufarg** [comment]
Yes the scheduler is on Redis. I believe since `agent is starting` is in the log, the agent should also be scheduled properly.

**Comment by aufarg** [comment]
So I managed to retrieve the thread dump for one of the agent when it's stuck. Apparently, the main agent execution is blocked when an agent tries to check if `customResourceDefinition` (might also affect other resouce) is readable in the cluster. I guess this is the same issue with what @clanesf describe above.

That being said, I think if a kind doesn't exist on the target cluster it shouldn't stuck indefinitely. In the thread dump, other agents' thread actually waiting in `TIMED_WAITING` state but in the case of stuck agent, it's waiting in `WAITING` state which I believe will wait indefinitely.

Last log from `AgentExecutionAction-1`:
```
<account name>/KubernetesUnregisteredCustomResourceCachingAgent[1/1]: agent is starting
Checking if customResourceDefinition is readable in account <account name>...
```

Last log from `AgentExecutionAction-4`:
```
<account name>/KubernetesCoreCachingAgent[1/1]: agent is starting
Checking if apiService is readable in account '<account name>'...
Kind apiService will not be cached in account '<account name>' because it cannot be listed.
Checking if clusterRole is readable in account '<account name>'...
Kind clusterRole will not be cached in account '<account name>' because it cannot be listed.
Checking if clusterRoleBinding is readable in account '<account name>'...
Kind clusterRoleBinding will not be cached in account '<account name>' because it cannot be listed.
Checking if configMap is readable in account '<account name>'...
Checking if controllerRevision is readable in account '<account name>'...
```

Thread dump from `AgentExecutionAction-1` and `AgentExecutionAction-4`:
```
"AgentExecutionAction-4" - Thread t@80
   java.lang.Thread.State: BLOCKED
        at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1674)
        - waiting to lock <1ddd1278> (a java.util.concurrent.ConcurrentHashMap$ReservationNode) owned by "AgentExecutionAction-1" t@77
        at com.netflix.spinnaker.clouddriver.kubernetes.security.KubernetesCredentials$PermissionValidator.isKindReadable(KubernetesCredentials.java:714)
        at com.netflix.spinnaker.clouddriver.kubernetes.security.KubernetesCredentials.getKindStatus(KubernetesCredentials.java:274)
        at com.netflix.spinnaker.clouddriver.kubernetes.security.KubernetesCredentials.isValidKind(KubernetesCredentials.java:246)
        at com.netflix.spinnaker.clouddriver.kubernetes.security.KubernetesCredentials$$Lambda$1200/99864940.test(Unknown Source)
        at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:174)
        at com.google.common.collect.CollectSpliterators$1WithCharacteristics.lambda$forEachRemaining$1(CollectSpliterators.java:67)
        at com.google.common.collect.CollectSpliterators$1WithCharacteristics$$Lambda$743/163766295.accept(Unknown Source)
        at java.util.stream.Streams$RangeIntSpliterator.forEachRemaining(Streams.java:110)
        at com.google.common.collect.CollectSpliterators$1WithCharacteristics.forEachRemaining(CollectSpliterators.java:67)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
        at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:566)
        at com.netflix.spinnaker.clouddriver.kubernetes.security.KubernetesCredentials.getGlobalKinds(KubernetesCredentials.java:290)
        at com.netflix.spinnaker.clouddriver.kubernetes.caching.agent.KubernetesCoreCachingAgent.primaryKinds(KubernetesCoreCachingAgent.java:59)
        at com.netflix.spinnaker.clouddriver.kubernetes.caching.agent.KubernetesV2CachingAgent.defaultIntrospectionDetails(KubernetesV2CachingAgent.java:92)
        at com.netflix.spinnaker.clouddriver.kubernetes.caching.agent.KubernetesV2OnDemandCachingAgent.loadData(KubernetesV2OnDemandCachingAgent.java:78)
        at com.netflix.spinnaker.cats.agent.CachingAgent$CacheExecution.executeAgentWithoutStore(CachingAgent.java:87)
        at com.netflix.spinnaker.cats.agent.CachingAgent$CacheExecution.executeAgent(CachingAgent.java:77)
        at com.netflix.spinnaker.cats.redis.cluster.ClusteredAgentScheduler$AgentExecutionAction.execute(ClusteredAgentScheduler.java:338)
        at com.netflix.spinnaker.cats.redis.cluster.ClusteredAgentScheduler$AgentJob.run(ClusteredAgentScheduler.java:308)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)

   Locked ownable synchronizers:
        - locked <588a317d> (a java.util.concurrent.ThreadPoolExecutor$Worker)

"AgentExecutionAction-1" - Thread t@77
   java.lang.Thread.State: WAITING
        at sun.misc.Unsafe.park(Native Method)
        - parking to wait for <198158bd> (a java.util.concurrent.FutureTask)
        at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
        at java.util.concurrent.FutureTask.awaitDone(FutureTask.java:429)
        at java.util.concurrent.FutureTask.get(FutureTask.java:191)
        at com.netflix.spinnaker.clouddriver.jobs.local.JobExecutorLocal.executeStreaming(JobExecutorLocal.java:112)
        at com.netflix.spinnaker.clouddriver.jobs.local.JobExecutorLocal.lambda$runJob$0(JobExecutorLocal.java:56)
        at com.netflix.spinnaker.clouddriver.jobs.local.JobExecutorLocal$$Lambda$1006/724061489.execute(Unknown Source)
        at com.netflix.spinnaker.clouddriver.jobs.local.JobExecutorLocal.executeWrapper(JobExecutorLocal.java:66)
        at com.netflix.spinnaker.clouddriver.jobs.local.JobExecutorLocal.runJob(JobExecutorLocal.java:56)
        at com.netflix.spinnaker.clouddriver.kubernetes.op.job.KubectlJobExecutor.list(KubectlJobExecutor.java:405)
        at com.netflix.spinnaker.clouddriver.kubernetes.security.KubernetesCredentials.lambda$list$5(KubernetesCredentials.java:396)
        at com.netflix.spinnaker.clouddriver.kubernetes.security.KubernetesCredentials$$Lambda$1020/866558640.get(Unknown Source)
        at com.netflix.spinnaker.clouddriver.kubernetes.security.KubernetesCredentials.runAndRecordMetrics(KubernetesCredentials.java:612)
        at com.netflix.spinnaker.clouddriver.kubernetes.security.KubernetesCredentials.runAndRecordMetrics(KubernetesCredentials.java:597)
        at com.netflix.spinnaker.clouddriver.kubernetes.security.KubernetesCredentials.list(KubernetesCredentials.java:391)
        at com.netflix.spinnaker.clouddriver.kubernetes.security.KubernetesCredentials$PermissionValidator.canReadKind(KubernetesCredentials.java:678)
        at com.netflix.spinnaker.clouddriver.kubernetes.security.KubernetesCredentials$PermissionValidator$$Lambda$1018/1503229074.apply(Unknown Source)
        at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1660)
        - locked <1ddd1278> (a java.util.concurrent.ConcurrentHashMap$ReservationNode)
        at com.netflix.spinnaker.clouddriver.kubernetes.security.KubernetesCredentials$PermissionValidator.isKindReadable(KubernetesCredentials.java:714)
        at com.netflix.spinnaker.clouddriver.kubernetes.security.KubernetesCredentials.getKindStatus(KubernetesCredentials.java:274)
        at com.netflix.spinnaker.clouddriver.kubernetes.security.KubernetesCredentials.isValidKind(KubernetesCredentials.java:246)
        at com.netflix.spinnaker.clouddriver.kubernetes.security.KubernetesCredentials.crdSupplier(KubernetesCredentials.java:306)
        at com.netflix.spinnaker.clouddriver.kubernetes.security.KubernetesCredentials$$Lambda$773/572892443.get(Unknown Source)
        at com.google.common.base.Suppliers$ExpiringMemoizingSupplier.get(Suppliers.java:243)
        - locked <aca52ee> (a com.google.common.base.Suppliers$ExpiringMemoizingSupplier)
        at com.netflix.spinnaker.clouddriver.kubernetes.security.KubernetesCredentials.getCrds(KubernetesCredentials.java:300)
        at com.netflix.spinnaker.clouddriver.kubernetes.caching.agent.KubernetesUnregisteredCustomResourceCachingAgent.primaryKinds(KubernetesUnregisteredCustomResourceCachingAgent.java:53)
        at com.netflix.spinnaker.clouddriver.kubernetes.caching.agent.KubernetesUnregisteredCustomResourceCachingAgent.primaryKinds(KubernetesUnregisteredCustomResourceCachingAgent.java:32)
        at com.netflix.spinnaker.clouddriver.kubernetes.caching.agent.KubernetesV2CachingAgent.defaultIntrospectionDetails(KubernetesV2CachingAgent.java:92)
        at com.netflix.spinnaker.clouddriver.kubernetes.caching.agent.KubernetesV2OnDemandCachingAgent.loadData(KubernetesV2OnDemandCachingAgent.java:78)
        at com.netflix.spinnaker.cats.agent.CachingAgent$CacheExecution.executeAgentWithoutStore(CachingAgent.java:87)
        at com.netflix.spinnaker.cats.agent.CachingAgent$CacheExecution.executeAgent(CachingAgent.java:77)
        at com.netflix.spinnaker.cats.redis.cluster.ClusteredAgentScheduler$AgentExecutionAction.execute(ClusteredAgentScheduler.java:338)
        at com.netflix.spinnaker.cats.redis.cluster.ClusteredAgentScheduler$AgentJob.run(ClusteredAgentScheduler.java:308)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)

   Locked ownable synchronizers:
        - locked <2b66fe33> (a java.util.concurrent.ThreadPoolExecutor$Worker)
```

**Comment by aufarg** [comment]
So the job executor stuck when waiting for output from `kubectl`. I haven't looked into the actual reason why `awaitDone` keeps waiting on parked object, most likely because somehow the stdout pipe is never closed so the task is also keeps waiting for EOF.

In the meantime, I think we should also introduce a timeout for local job executor so agents are not stuck indefinitely. I made a PR that does that: https://github.com/spinnaker/clouddriver/pull/5201

**Comment by abhilashjoseph** [comment]
We have a fairly crowded cluster, with multiple namespaces have different types of services running in it, adding more than one account of siimilar clusters seemed to increase the cpu usage considerably. The reason we believe is because the kubernetes caching agents default polling interval (30s) is low, which results in initiating caching more frequently than it completely, we mitigated it by changing the following setting for the cloud driver
```
sql:
  agent:
    poll:
      intervalSeconds: 1800
```

**Comment by aufarg** [comment]
@abhilashjoseph based on the thread dump it's stuck when trying to run `kubectl`. It's a little bit proven when I tried to apply my PR above and we haven't faced with this issue (the exception is also thrown multiple times). As for why why `kubectl` stuck, I can't really say.

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by hisener** [comment]
@spinnakerbot remove-label stale

I am also having the same problem in 1.24.4.

**Comment by rampatnaik** [comment]
hello 

any update for this PR

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by hisener** [comment]
@spinnakerbot remove-label stale

**Comment by spinnakerbot** [comment]
This issue hasn't been updated in 45 days, so we are tagging it as 'stale'. If you want to remove this label, comment:

> @spinnakerbot remove-label stale

**Comment by hisener** [comment]
@spinnakerbot remove-label stale
